{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Targeted Cluster Analysis using cBLASTER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import altair as alt\n",
    "import glob\n",
    "from Bio import Phylo\n",
    "from Bio import SeqIO\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Proejct initiation stage\n",
    "Define project name, rest will be created automatically when code is exicuted in the working directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Define project name'''\n",
    "project_name=\"Streptomyces_Azoxy_Distribution\"\n",
    "'''Define location for cblaster database'''\n",
    "dir_cblaster_db = \"custom_db/\"+\"NCBI_P8A2_20230216/\"\n",
    "\n",
    "'''Defult directory layout'''\n",
    "input_dir=\"input/\"+project_name+\"/\"\n",
    "output_dir=\"output/\"+project_name+\"/\"\n",
    "dir_clinker = output_dir+\"clinker/\"\n",
    "dir_cblaster = output_dir+\"cblaster/\"\n",
    "dir_iTOL = output_dir+\"iTOL/\"\n",
    "\n",
    "'''Automated creation of directories'''\n",
    "directories = (input_dir,output_dir, dir_clinker, dir_cblaster, dir_iTOL)\n",
    "'''Check whether the specified path exists or not. If dir doesnt exist, it is created'''\n",
    "for path in directories:\n",
    "   isExist = os.path.exists(path)\n",
    "   if not isExist:\n",
    "      # Create a new directory because it does not exist\n",
    "      os.makedirs(path)\n",
    "      print(\"Directory is created - \"+path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insert relevent files into the project input directory\n",
    "Following files are accepted:\n",
    "- Files containing complete gene cluster which will be passed through clinker to find gene pairs: ***.gbk** and/or ***.gb** format\n",
    "- any protein fasta file that should be included in analysis without gene grouping, must contain only cluster specific genes as they will be latter added to the list of core/tailoring genes as singletons: protein ***.fasta** format.\n",
    "- List of protein IDs that have been characterised, list will be created automatically from fasta files in the folder, or can be done manually. The list will be checked against clinker created gene groups and cluster specific gene ids will be added to the list of groups --> **characterised_protein_IDs.csv** file\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Definitions used inside the project, generated by chatGPT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def export_protein_ids(gbk_file):\n",
    "    protein_ids = []\n",
    "\n",
    "    # Read the GenBank file and extract protein IDs\n",
    "    with open(gbk_file, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            for feature in record.features:\n",
    "                if feature.type == \"CDS\" and \"protein_id\" in feature.qualifiers:\n",
    "                    protein_ids.append(feature.qualifiers[\"protein_id\"][0])\n",
    "\n",
    "    return protein_ids\n",
    "\n",
    "def genbank_to_fasta(input_file, output_file):\n",
    "    # Open the input GenBank file\n",
    "    with open(input_file, \"r\") as handle:\n",
    "        # Parse the GenBank file and store the records\n",
    "        records = list(SeqIO.parse(handle, \"genbank\"))\n",
    "\n",
    "    # Open the output FASTA file\n",
    "    with open(output_file, \"w\") as handle:\n",
    "        # Iterate over the records and write protein sequences to the FASTA file\n",
    "        for record in records:\n",
    "            for feature in record.features:\n",
    "                if feature.type == \"CDS\":\n",
    "                    if \"translation\" in feature.qualifiers:\n",
    "                        protein_sequence = feature.qualifiers[\"translation\"][0]\n",
    "                        handle.write(f\">{feature.qualifiers['protein_id'][0]}\\n\")\n",
    "                        handle.write(f\"{protein_sequence}\\n\")\n",
    "\n",
    "def rgb_to_hex(rgb):\n",
    "    # Validate RGB values\n",
    "    if len(rgb) != 3:\n",
    "        raise ValueError(\"RGB values must be a list or tuple of length 3.\")\n",
    "    # Convert RGB to hexadecimal\n",
    "    hex_color = '#{:02x}{:02x}{:02x}'.format(*rgb)\n",
    "    return hex_color\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creation of gene groups using clinker and populating it using list of cluster specific genes eather as protein.fasta file or list of protein IDs in provided .csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory is not empty. Skipping the step.\n"
     ]
    }
   ],
   "source": [
    "'''Running clinker on *.gb and *.gbk files in the project input directory. Outputs are generated in clinker folder located in output project directory.'''\n",
    "if not os.listdir(dir_clinker):\n",
    "    os.environ[\"dir_clinker\"] = dir_clinker\n",
    "    os.environ[\"input_dir\"] = input_dir\n",
    "    !bash shell_script_clinker.sh\n",
    "else:\n",
    "    print(\"Directory is not empty. Skipping the step.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'AAN10236.1', 'PRJNA935111:PU648_09280', 'WP_037654301.1'},\n",
       " {'AAN10237.1', 'PRJNA935111:PU648_09235', 'WP_037654320.1'},\n",
       " {'AAN10238.1', 'PRJNA935111:PU648_09285', 'WP_037654315.1'},\n",
       " {'AAN10239.1', 'PRJNA935111:PU648_09255', 'WP_154822324.1'},\n",
       " {'AAN10240.1', 'PRJNA935111:PU648_09275', 'WP_037654313.1'},\n",
       " {'AAN10241.1', 'WP_037654308.1'},\n",
       " {'AAN10246.1', 'WP_037654300.1'},\n",
       " {'AAN10247.1', 'WP_037654338.1'},\n",
       " {'AAN10248.1', 'WP_125207584.1'},\n",
       " {'AAN10249.1', 'WP_037654307.1'},\n",
       " {'PRJNA935111:PU648_09245', 'WP_037654318.1'},\n",
       " {'PRJNA935111:PU648_09250', 'WP_037654316.1'},\n",
       " {'PRJNA935111:PU648_09270', 'WP_051850081.1'}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Code written together with thom to create the gene matrix automatically\n",
    "df_genes_list = pd.read_csv(dir_clinker +'alignments.csv',header=None,usecols=[0,1]).dropna()\n",
    "df_genes_list.rename( columns={0 :'Pair_gene_1'}, inplace=True )\n",
    "df_genes_list.rename( columns={1 :'Pair_gene_2'}, inplace=True )\n",
    "# Create an empty list\n",
    "genes_list =[]\n",
    "# Iterate over each row\n",
    "for rows in df_genes_list.itertuples():\n",
    "    # Create list for the current row\n",
    "    each_pair =[rows.Pair_gene_1, rows.Pair_gene_2]\n",
    "    # append the list to the final list\n",
    "    genes_list.append(each_pair)\n",
    "\n",
    "gene_group_list = [{genes_list[0][0]}]\n",
    "for genes in genes_list:\n",
    "    genes_added = False\n",
    "    for gene_group in gene_group_list:\n",
    "        if genes[0] in gene_group or genes[1] in gene_group:\n",
    "            gene_group.add(genes[1])\n",
    "            gene_group.add(genes[0])\n",
    "            genes_added = True\n",
    "    if genes_added == False:\n",
    "        gene_group_list.extend([{genes[0]}])\n",
    "        for gene_group in gene_group_list:\n",
    "            if genes[0] in gene_group or genes[1] in gene_group:\n",
    "                gene_group.add(genes[1])\n",
    "                gene_group.add(genes[0])\n",
    "sorted(gene_group_list)\n",
    "gene_group_list.sort(key = len, reverse=True)\n",
    "\n",
    "gene_group_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Adding singletons to the generated pairs from fasta file and custom .csv list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Addition of known cluster specific tailoring genes to the gene list is done here, these genes are essential for specific cluster that are not shared between the clusters'''\n",
    "'''The protein IDs are taken from protein fasta file of characterized enzymes for each BGC, list must be provided manually as input'''\n",
    "\n",
    "'''firstly we translate the protein_fasta_file into csv list of protein IDs'''\n",
    "'''Note: this can be skipped if file is manually populated with IDs'''\n",
    "# Create an empty list to store sequence names from fasta file\n",
    "protein_sequence_ids = []\n",
    "\n",
    "# Loop through each fasta file in the directory\n",
    "for filename in os.listdir(input_dir):\n",
    "    if filename.endswith(\".fasta\"):\n",
    "        # Open and read the current FASTA file\n",
    "        with open(os.path.join(input_dir, filename), \"r\") as fasta_file:\n",
    "            for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "                # Extract sequence name and add to the list\n",
    "                protein_sequence_id = record.id\n",
    "                protein_sequence_ids.append(protein_sequence_id)\n",
    "\n",
    "'''Add protein IDs from the fasta file to predefined core genes'''\n",
    "df_fasta_protein_IDs = pd.DataFrame({\"protein_id\": protein_sequence_ids})\n",
    "df__manual_protein_IDs = pd.read_csv(input_dir+\"characterised_protein_IDs.csv\")\n",
    "\n",
    "df_protein_ids_merged = pd.merge(df_fasta_protein_IDs, df__manual_protein_IDs, on='protein_id', how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overview of protein pairs\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'AAN10236.1', 'PRJNA935111:PU648_09280', 'WP_037654301.1'},\n",
       " {'AAN10237.1', 'PRJNA935111:PU648_09235', 'WP_037654320.1'},\n",
       " {'AAN10238.1', 'PRJNA935111:PU648_09285', 'WP_037654315.1'},\n",
       " {'AAN10239.1', 'PRJNA935111:PU648_09255', 'WP_154822324.1'},\n",
       " {'AAN10240.1', 'PRJNA935111:PU648_09275', 'WP_037654313.1'},\n",
       " {'AAN10241.1', 'WP_037654308.1'},\n",
       " {'AAN10246.1', 'WP_037654300.1'},\n",
       " {'AAN10247.1', 'WP_037654338.1'},\n",
       " {'AAN10248.1', 'WP_125207584.1'},\n",
       " {'AAN10249.1', 'WP_037654307.1'},\n",
       " {'PRJNA935111:PU648_09245', 'WP_037654318.1'},\n",
       " {'PRJNA935111:PU648_09250', 'WP_037654316.1'},\n",
       " {'PRJNA935111:PU648_09270', 'WP_051850081.1'},\n",
       " {'KP687747'},\n",
       " {'KP687746'},\n",
       " {'KP687745'},\n",
       " {'KP687744'},\n",
       " {'KP687743'},\n",
       " {'KP687742'},\n",
       " {'KP687741'},\n",
       " {'KP687740'},\n",
       " {'KP687739'},\n",
       " {'KP687738'},\n",
       " {'KP687737'},\n",
       " {'KP687736'},\n",
       " {'KP687735'},\n",
       " {'KP687734'},\n",
       " {'KP687733'},\n",
       " {'AAN10242.1'},\n",
       " {'AAN10243.1'},\n",
       " {'AAN10244.1'},\n",
       " {'PRJNA935111:PU648_09290'},\n",
       " {'PRJNA935111:PU648_09240'},\n",
       " {'PRJNA935111:PU648_09260'},\n",
       " {'PRJNA935111:PU648_09325'},\n",
       " {'PRJNA935111:PU648_09360'}]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''Filtering of singelton sets from clinker detected pairs and adding unique ones from df_protein_ids_merged to the gene list'''\n",
    "# Initialize an empty list to store the singleton set\n",
    "gene_singleton_list_of_sets = []\n",
    "# Iterate through the sequence names in the DataFrame and create singleton set\n",
    "for protein_id in df_protein_ids_merged[\"protein_id\"]:\n",
    "    singleton_set = {protein_id}\n",
    "    gene_singleton_list_of_sets.append(singleton_set)\n",
    "\n",
    "'''make clinker detected gene pair list of sets into single set list to filter out singleton set from existing values'''\n",
    "clinker_protein_id_set = set()\n",
    "for group in gene_group_list:\n",
    "    for gene in group:\n",
    "        clinker_protein_id_set.add(gene)\n",
    "\n",
    "'''filtering of singelton set and joining them to clinker pairs'''\n",
    "for singelton_set in gene_singleton_list_of_sets:\n",
    "    filtered_set = singelton_set.difference(clinker_protein_id_set)\n",
    "    if filtered_set:\n",
    "        gene_group_list.append(filtered_set)\n",
    "\n",
    "\n",
    "print(\"Overview of protein pairs\")\n",
    "gene_group_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes (which are contained by) number of groups\n",
      "23 1\n",
      "8 2\n",
      "5 3\n"
     ]
    }
   ],
   "source": [
    "'''To have better overview, for parameter estimate, the number of genes contained in each group is calculated below'''\n",
    "from collections import defaultdict\n",
    "counts = defaultdict(int)\n",
    "for main_group in gene_group_list:\n",
    "    counts[sum(gene in group for group in gene_group_list for gene in main_group)] += 1\n",
    "counts = dict(counts)\n",
    "\n",
    "print(\"Number of genes (which are contained by) number of groups\")\n",
    "for i in range(1, max(counts)+1):\n",
    "    print(counts.get(i, 0), i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parameter selection for filtering of cblaster results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dependencies:\n",
    "'''These variables are used for filtering of cblaster table based on list of core genes and list of tailoring genes shared between input clusters'''\n",
    "min_genes_shared_to_be_core = 3         # Largest number of genes in a group\n",
    "minimum_core_genes = 4                  # Set to 4 as max is 5, we are not sure if all 5 are required\n",
    "minimum_gene_hits = 5           # set to 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Target gene list:\n",
      "['AAN10236.1', 'AAN10237.1', 'AAN10238.1', 'AAN10239.1', 'AAN10240.1', 'AAN10241.1', 'AAN10242.1', 'AAN10243.1', 'AAN10244.1', 'AAN10246.1', 'AAN10247.1', 'AAN10248.1', 'AAN10249.1', 'KP687733', 'KP687734', 'KP687735', 'KP687736', 'KP687737', 'KP687738', 'KP687739', 'KP687740', 'KP687741', 'KP687742', 'KP687743', 'KP687744', 'KP687745', 'KP687746', 'KP687747', 'PRJNA935111:PU648_09235', 'PRJNA935111:PU648_09240', 'PRJNA935111:PU648_09245', 'PRJNA935111:PU648_09250', 'PRJNA935111:PU648_09255', 'PRJNA935111:PU648_09260', 'PRJNA935111:PU648_09270', 'PRJNA935111:PU648_09275', 'PRJNA935111:PU648_09280', 'PRJNA935111:PU648_09285', 'PRJNA935111:PU648_09290', 'PRJNA935111:PU648_09325', 'PRJNA935111:PU648_09360', 'WP_037654300.1', 'WP_037654301.1', 'WP_037654307.1', 'WP_037654308.1', 'WP_037654313.1', 'WP_037654315.1', 'WP_037654316.1', 'WP_037654318.1', 'WP_037654320.1', 'WP_037654338.1', 'WP_051850081.1', 'WP_125207584.1', 'WP_154822324.1']\n",
      "Core gene list:\n",
      "['AAN10236.1', 'AAN10237.1', 'AAN10238.1', 'AAN10239.1', 'AAN10240.1', 'PRJNA935111:PU648_09235', 'PRJNA935111:PU648_09255', 'PRJNA935111:PU648_09275', 'PRJNA935111:PU648_09280', 'PRJNA935111:PU648_09285', 'WP_037654301.1', 'WP_037654313.1', 'WP_037654315.1', 'WP_037654320.1', 'WP_154822324.1']\n",
      "Tailoring gene list:\n",
      "['AAN10241.1', 'AAN10242.1', 'AAN10243.1', 'AAN10244.1', 'AAN10246.1', 'AAN10247.1', 'AAN10248.1', 'AAN10249.1', 'KP687733', 'KP687734', 'KP687735', 'KP687736', 'KP687737', 'KP687738', 'KP687739', 'KP687740', 'KP687741', 'KP687742', 'KP687743', 'KP687744', 'KP687745', 'KP687746', 'KP687747', 'PRJNA935111:PU648_09240', 'PRJNA935111:PU648_09245', 'PRJNA935111:PU648_09250', 'PRJNA935111:PU648_09260', 'PRJNA935111:PU648_09270', 'PRJNA935111:PU648_09290', 'PRJNA935111:PU648_09325', 'PRJNA935111:PU648_09360', 'WP_037654300.1', 'WP_037654307.1', 'WP_037654308.1', 'WP_037654316.1', 'WP_037654318.1', 'WP_037654338.1', 'WP_051850081.1', 'WP_125207584.1']\n"
     ]
    }
   ],
   "source": [
    "'''Genes are classified in three groups, one group contains all of the rarget genes, the secound group contains only core genes and the final group contains all genes that are not core.'''\n",
    "target_gene_list = []\n",
    "core_genes_list = []\n",
    "tailoring_genes_list = []\n",
    "for group in gene_group_list:\n",
    "    for gene in group:\n",
    "            target_gene_list.append(gene)\n",
    "    if len(group) == min_genes_shared_to_be_core:\n",
    "        for core_gene in group:\n",
    "            core_genes_list.append(core_gene)\n",
    "    else:\n",
    "        for tailoring_gene in group:\n",
    "            tailoring_genes_list.append(tailoring_gene)\n",
    "target_gene_list.sort()\n",
    "core_genes_list.sort()\n",
    "tailoring_genes_list.sort()\n",
    "print(\"Target gene list:\")\n",
    "print(target_gene_list)\n",
    "print(\"Core gene list:\")\n",
    "print(core_genes_list)\n",
    "print(\"Tailoring gene list:\")\n",
    "print(tailoring_genes_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Azodyrecin': 'PRJNA935111:PU648_', 'Azoxymycin': 'KP', 'KA57A': 'WP_', 'Valanimycin': 'AAN'}\n"
     ]
    }
   ],
   "source": [
    "'''define prephase code of protein_IDs for each imput BGC manually'''\n",
    "\n",
    "BGC_protein_ID_codes_dict = {\n",
    "    'Azodyrecin': 'PRJNA935111:PU648_',\n",
    "    'Azoxymycin': 'KP',\n",
    "    'KA57A': 'WP_',\n",
    "    'Valanimycin': 'AAN'\n",
    "}\n",
    "\n",
    "# Printing the dictionary\n",
    "print(BGC_protein_ID_codes_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cblaster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''all of the input files are combined in one protein fasta file. Following code translates each of the GenBank files to fasta files and merge them into single input file for cblaster'''\n",
    "'''Translation of gbk to fasta files'''\n",
    "list_of_GenBank_files = []\n",
    "for file in os.listdir(input_dir):\n",
    "    if file.endswith(\".gb\") or file.endswith(\".gbk\"):\n",
    "        list_of_GenBank_files.append(os.path.join(input_dir,file))\n",
    "\n",
    "list_of_fasta_files = []\n",
    "for file in os.listdir(input_dir):\n",
    "    if file.endswith(\".gb\") or file.endswith(\".gbk\"):\n",
    "        file=file.replace('.gbk','.fasta')\n",
    "        file=file.replace('.gb','.fasta')\n",
    "        list_of_fasta_files.append(os.path.join(dir_cblaster,file))\n",
    "\n",
    "\n",
    "for genbank_input, fasta_output in zip(list_of_GenBank_files,list_of_fasta_files):\n",
    "    genbank_to_fasta(genbank_input, fasta_output)\n",
    "\n",
    "'''Merge fasta files into one input fasta file for cblaster'''\n",
    "for file in os.listdir(input_dir):\n",
    "    if file.endswith(\".fasta\"):\n",
    "        list_of_fasta_files.append(os.path.join(input_dir,file))\n",
    "list_of_fasta_files\n",
    "\n",
    "fasta_sequences = []\n",
    "for file in list_of_fasta_files:\n",
    "    with open(file, 'r') as fasta_file:\n",
    "        fasta_sequences.extend(SeqIO.parse(fasta_file, 'fasta'))\n",
    "\n",
    "merged_fasta_file = dir_cblaster+\"merged_fasta_file_cblaster_input.fasta\"\n",
    "with open(merged_fasta_file, 'w') as output_fasta:\n",
    "    SeqIO.write(fasta_sequences, output_fasta, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory contains cblaster cblaster_output.csv file. Skipping the step.\n"
     ]
    }
   ],
   "source": [
    "'''running cblaster'''\n",
    "if not os.path.exists(dir_cblaster+\"cblaster_output.csv\"):\n",
    "    '''parameter import from notebook to bash enviroment'''\n",
    "    os.environ[\"input_fasta\"] = merged_fasta_file\n",
    "    os.environ[\"output\"] = dir_cblaster+\"cblaster_output\"\n",
    "    os.environ[\"minimum_hits\"] = str(minimum_gene_hits)\n",
    "    os.environ[\"cblaster_db\"] = dir_cblaster_db\n",
    "    '''cblaster run'''\n",
    "    !bash shell_script_cblaster.sh\n",
    "else:\n",
    "    print('Directory contains cblaster cblaster_output.csv file. Skipping the step.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering of cblaster output\n",
    "Firstly we reduce table to contain only marked genes in the **gene_group_list** and we keep only the top result for homologous gene hits from different BGCs, keeping only the cloesest identity match.\n",
    "\n",
    "Then we assign which BGC the cluster belongs to based on previously manually defined **BGC_protein_ID_codes_dict**\n",
    "\n",
    "Finally we do filtering of the table by firstly counting up the gene and core gene hits and then filtering it based on the definitions entered in **minimum_core_genes** and **minimum_gene_hits**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Define BGCs that are to be excluded from core gene requirement filtering, meaning they dont have hits in clinker'''\n",
    "excluded_from_core_req_dict = {\n",
    "    'Azodyrecin': 'include',\n",
    "    'Azoxymycin': 'exclude',\n",
    "    'KA57A': 'include',\n",
    "    'Valanimycin': 'include'\n",
    "}\n",
    "\n",
    "'''define list of genes that must be contained in BGCs excluded from core filtering'''\n",
    "custom_required_genes_for_non_core_filt = [\n",
    "    'KP687735'\n",
    "]\n",
    "number_of_req_custom_genes = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1926/1870334211.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_tmp['count'] = df_tmp[custom_required_genes_for_non_core_filt].count(axis=1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organism</th>\n",
       "      <th>Scaffold</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>AAN10236.1</th>\n",
       "      <th>AAN10237.1</th>\n",
       "      <th>AAN10238.1</th>\n",
       "      <th>AAN10239.1</th>\n",
       "      <th>AAN10240.1</th>\n",
       "      <th>...</th>\n",
       "      <th>Azodyrecin gene hits</th>\n",
       "      <th>Azodyrecin %input</th>\n",
       "      <th>Azoxymycin gene hits</th>\n",
       "      <th>Azoxymycin %input</th>\n",
       "      <th>KA57A gene hits</th>\n",
       "      <th>KA57A %input</th>\n",
       "      <th>Valanimycin gene hits</th>\n",
       "      <th>Valanimycin %input</th>\n",
       "      <th>bgc</th>\n",
       "      <th>core gene hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_012922115.1</td>\n",
       "      <td>JABAQG010000001</td>\n",
       "      <td>10106302</td>\n",
       "      <td>10179481</td>\n",
       "      <td>74.1437</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Azodyrecin</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_008704445.1</td>\n",
       "      <td>NZ_CP023697.1</td>\n",
       "      <td>2399106</td>\n",
       "      <td>2443830</td>\n",
       "      <td>37.0286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>1</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>KA57A</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_008705135.1</td>\n",
       "      <td>NZ_CP023694.1</td>\n",
       "      <td>6954730</td>\n",
       "      <td>6992504</td>\n",
       "      <td>57.7660</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>KA57A</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_014650675.1</td>\n",
       "      <td>NZ_BMVK01000003</td>\n",
       "      <td>343098</td>\n",
       "      <td>384521</td>\n",
       "      <td>40.0731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>10</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>1</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>KA57A</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_003112515.1</td>\n",
       "      <td>NZ_BEVZ01000002</td>\n",
       "      <td>430776</td>\n",
       "      <td>447432</td>\n",
       "      <td>27.6654</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>1</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>KA57A</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>GCF_016876065.1</td>\n",
       "      <td>c00001_NZ_JABZ..</td>\n",
       "      <td>1034874</td>\n",
       "      <td>1047113</td>\n",
       "      <td>24.6673</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Azoxymycin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>GCF_014655715.1</td>\n",
       "      <td>NZ_BNBM01000001</td>\n",
       "      <td>1175745</td>\n",
       "      <td>1203475</td>\n",
       "      <td>12.3384</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Azoxymycin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>GCF_002797835.1</td>\n",
       "      <td>NZ_PGEO01000001</td>\n",
       "      <td>6294577</td>\n",
       "      <td>6307152</td>\n",
       "      <td>26.6909</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Azoxymycin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>GCF_003935055.1</td>\n",
       "      <td>NZ_RHGC01000006</td>\n",
       "      <td>12415</td>\n",
       "      <td>75424</td>\n",
       "      <td>12.3326</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>1</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Azoxymycin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>GCF_021173675.1</td>\n",
       "      <td>c00038_NZ_JAIH..</td>\n",
       "      <td>520238</td>\n",
       "      <td>534632</td>\n",
       "      <td>9.2364</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Azoxymycin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>375 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Organism          Scaffold     Start       End    Score  \\\n",
       "0    GCF_012922115.1   JABAQG010000001  10106302  10179481  74.1437   \n",
       "1    GCF_008704445.1     NZ_CP023697.1   2399106   2443830  37.0286   \n",
       "2    GCF_008705135.1     NZ_CP023694.1   6954730   6992504  57.7660   \n",
       "3    GCF_014650675.1   NZ_BMVK01000003    343098    384521  40.0731   \n",
       "4    GCF_003112515.1   NZ_BEVZ01000002    430776    447432  27.6654   \n",
       "..               ...               ...       ...       ...      ...   \n",
       "392  GCF_016876065.1  c00001_NZ_JABZ..   1034874   1047113  24.6673   \n",
       "394  GCF_014655715.1   NZ_BNBM01000001   1175745   1203475  12.3384   \n",
       "395  GCF_002797835.1   NZ_PGEO01000001   6294577   6307152  26.6909   \n",
       "396  GCF_003935055.1   NZ_RHGC01000006     12415     75424  12.3326   \n",
       "402  GCF_021173675.1  c00038_NZ_JAIH..    520238    534632   9.2364   \n",
       "\n",
       "     AAN10236.1  AAN10237.1  AAN10238.1  AAN10239.1  AAN10240.1  ...  \\\n",
       "0           NaN         NaN         NaN         NaN         NaN  ...   \n",
       "1           NaN         NaN         NaN         NaN         NaN  ...   \n",
       "2           NaN         NaN         NaN         NaN         NaN  ...   \n",
       "3           NaN         NaN         NaN         NaN         NaN  ...   \n",
       "4           NaN         NaN         NaN         NaN         NaN  ...   \n",
       "..          ...         ...         ...         ...         ...  ...   \n",
       "392         NaN         NaN         NaN         NaN         NaN  ...   \n",
       "394         NaN         NaN         NaN         NaN         NaN  ...   \n",
       "395         NaN         NaN         NaN         NaN         NaN  ...   \n",
       "396         NaN         NaN         NaN         NaN         NaN  ...   \n",
       "402         NaN         NaN         NaN         NaN         NaN  ...   \n",
       "\n",
       "     Azodyrecin gene hits  Azodyrecin %input  Azoxymycin gene hits  \\\n",
       "0                      13                1.0                     0   \n",
       "1                       0                0.0                     0   \n",
       "2                       0                0.0                     0   \n",
       "3                       0                0.0                     1   \n",
       "4                       0                0.0                     0   \n",
       "..                    ...                ...                   ...   \n",
       "392                     0                0.0                    14   \n",
       "394                     0                0.0                     2   \n",
       "395                     0                0.0                    15   \n",
       "396                     0                0.0                     4   \n",
       "402                     0                0.0                     6   \n",
       "\n",
       "     Azoxymycin %input  KA57A gene hits  KA57A %input  Valanimycin gene hits  \\\n",
       "0             0.000000                0      0.000000                      0   \n",
       "1             0.000000               11      0.846154                      1   \n",
       "2             0.000000               12      0.923077                      0   \n",
       "3             0.066667               10      0.769231                      1   \n",
       "4             0.000000               10      0.769231                      1   \n",
       "..                 ...              ...           ...                    ...   \n",
       "392           0.933333                0      0.000000                      0   \n",
       "394           0.133333                0      0.000000                      0   \n",
       "395           1.000000                0      0.000000                      0   \n",
       "396           0.266667                1      0.076923                      0   \n",
       "402           0.400000                0      0.000000                      0   \n",
       "\n",
       "     Valanimycin %input         bgc  core gene hits  \n",
       "0              0.000000  Azodyrecin               5  \n",
       "1              0.076923       KA57A               5  \n",
       "2              0.000000       KA57A               5  \n",
       "3              0.076923       KA57A               5  \n",
       "4              0.076923       KA57A               5  \n",
       "..                  ...         ...             ...  \n",
       "392            0.000000  Azoxymycin               0  \n",
       "394            0.000000  Azoxymycin               0  \n",
       "395            0.000000  Azoxymycin               0  \n",
       "396            0.000000  Azoxymycin               0  \n",
       "402            0.000000  Azoxymycin               0  \n",
       "\n",
       "[375 rows x 70 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''load in cblaster output'''\n",
    "cblaster_output = pd.read_csv(dir_cblaster+\"cblaster_output.csv\")\n",
    "\n",
    "'''make blank pandas dataframe containing only gene group lists'''\n",
    "columns_to_keep = list(cblaster_output)[:5] + target_gene_list\n",
    "columns_to_drop = [col for col in cblaster_output.columns if col not in columns_to_keep]\n",
    "df_cblaster = cblaster_output.drop(columns=columns_to_drop)\n",
    "df_cblaster\n",
    "\n",
    "'''remove columns that do not have sufficient number of gene hits from target_gene_list'''\n",
    "df_cblaster['gene hits'] = df_cblaster[target_gene_list].count(axis=1)\n",
    "df_cblaster = df_cblaster[df_cblaster['gene hits'] >= minimum_gene_hits]\n",
    "\n",
    "for group in gene_group_list:\n",
    "    if len(group) >1:\n",
    "        group_list = list(group)\n",
    "        max_values = df_cblaster[group_list].max(axis=1)\n",
    "        for protein_id in group:\n",
    "            df_cblaster.loc[df_cblaster[protein_id] != max_values, protein_id] = np.nan\n",
    "df_cblaster = df_cblaster.replace(0, np.nan)\n",
    "\n",
    "for bgc, protein_ID_key in BGC_protein_ID_codes_dict.items():\n",
    "    relevant_columns = [column for column in df_cblaster.columns if column.startswith(protein_ID_key)]\n",
    "    df_cblaster[bgc+' gene hits'] = df_cblaster[relevant_columns].count(axis=1)\n",
    "    df_cblaster[bgc+' %input'] = df_cblaster[bgc+' gene hits']/len(relevant_columns)\n",
    "\n",
    "bgc_list = []\n",
    "for bgc in BGC_protein_ID_codes_dict:\n",
    "    bgc_list.append(bgc+' %input')\n",
    "df_cblaster['bgc'] = df_cblaster[bgc_list].idxmax(axis=1)\n",
    "df_cblaster['bgc'] = df_cblaster['bgc'].str.replace(' %input', '')\n",
    "\n",
    "'''count number of core genes'''\n",
    "df_cblaster['core gene hits'] = df_cblaster[core_genes_list].count(axis=1)\n",
    "'''Removing genomes with insuficient core gene hits from list of genomes that should be filrered, except if bgc is excepted and alternative rule paplies'''\n",
    "for bgc, dont_perform_core_filtering in excluded_from_core_req_dict.items():\n",
    "    if dont_perform_core_filtering == 'exclude':\n",
    "        df_tmp = df_cblaster[df_cblaster['bgc'] == bgc]\n",
    "        df_tmp['count'] = df_tmp[custom_required_genes_for_non_core_filt].count(axis=1)\n",
    "        df_tmp = df_tmp[df_tmp['count'] < number_of_req_custom_genes]\n",
    "        df_cblaster = df_cblaster.drop(df_tmp.index)\n",
    "    else:\n",
    "        df_tmp = df_cblaster[df_cblaster['bgc'] == bgc]\n",
    "        df_tmp = df_tmp[df_tmp['core gene hits'] < minimum_core_genes]\n",
    "        df_cblaster = df_cblaster.drop(df_tmp.index)\n",
    "\n",
    "\n",
    "'''Sorting of table'''\n",
    "df_cblaster = df_cblaster.sort_values(by ='Organism', ascending = True)\n",
    "df_cblaster = df_cblaster.sort_values(by ='Scaffold', ascending = True)\n",
    "df_cblaster = df_cblaster.sort_values(by ='gene hits', ascending = False)\n",
    "df_cblaster = df_cblaster.sort_values(by ='core gene hits', ascending = False)\n",
    "df_cblaster = df_cblaster.reset_index(drop=True)\n",
    "\n",
    "'''As for iTOL visualization only 1 cluster can be showed for single node,\n",
    "we exclude worst hit for genomes with two or more hits and save it in seperate file'''\n",
    "duplicated_mask = df_cblaster.duplicated('Organism', keep='first')\n",
    "df_cblaster_excluded = df_cblaster[duplicated_mask]\n",
    "df_cblaster = df_cblaster[~duplicated_mask]\n",
    "'''saving tables to *.csv files'''\n",
    "df_cblaster_excluded.to_csv(dir_cblaster+\"df_cblaster_excluded.csv\",index=False)\n",
    "df_cblaster.to_csv(dir_cblaster+\"df_cblaster.csv\",index=False)\n",
    "\n",
    "'''Creating additional table where gene rows are grouped in specific groups that will be used later in iTOL visualizations'''\n",
    "#not done for now\n",
    "df_cblaster"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iTOL visualization and splitting data into sub-tables\n",
    "- To visalize the data, create a table where Each cluster is given its own color\n",
    "    - Group the most similar clusters with XX cutoff to be responsible for production of best mach cBLASTER cluster\n",
    "        - Probably best to use max cluster hits / with core gene hits to get estimated ratio.\n",
    "            - This can create problems if input cluster is not part of the diamond database and better would be to use clinker identifited gene coutn. Requires to make table with cblaster_input and clinker hits\n",
    "    - Low similarity clusters are offcolored to be something similar to target input\n",
    "    - very-low similarity clusters are graphed as other related clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Azodyrecin': '204, 121, 167',\n",
       " 'Azoxymycin': '0, 158, 115',\n",
       " 'KA57A': '86, 180, 233',\n",
       " 'Valanimycin': '213, 94, 0'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BGC_RBG_color_dict = {\n",
    "    'Azodyrecin': '204, 121, 167',\n",
    "    'Azoxymycin': '0, 158, 115',\n",
    "    'KA57A': '86, 180, 233',\n",
    "    'Valanimycin': '213, 94, 0'\n",
    "}\n",
    "BGC_RBG_color_dict"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## color_gradient/color_ranges\n",
    "The resulting table can be used in iTOL to copy-paste data from excel file into **iTOL color gradiant class**, or **color ranges** which can only be eddited automatically using **iTOL annotation edditor** requiering extra subscription in iTOL.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create iTOL annotation table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Creation of df to be populated with data'''\n",
    "df_iTOL_cg = pd.DataFrame(columns=['Tree node (bgcflow input file name)','Color (rgba(255, 255, 255, 1.000))','Label'])\n",
    "'''Populate the table with df_cblaster results'''\n",
    "df_iTOL_cg['Tree node (bgcflow input file name)'] = df_cblaster['Organism'] \n",
    "df_iTOL_cg['Label'] = df_cblaster['bgc']\n",
    "'''color intensity is the % of genes hits belonging to flagged bgc which is used to change intensity of rgb color'''\n",
    "color_intensity = df_cblaster[bgc_list].max(axis=1)\n",
    "#for bgc, rgb in BGC_RBG_color_dict.items():\n",
    "#    if df_iTOL_cg['Label'] == bgc:\n",
    "df_iTOL_cg['Color (rgba(255, 255, 255, 1.000))'] = df_iTOL_cg['Label'].map(BGC_RBG_color_dict)\n",
    "df_iTOL_cg['Color (rgba(255, 255, 255, 1.000))'] = 'rgba(' + df_iTOL_cg['Color (rgba(255, 255, 255, 1.000))'] + ', ' + color_intensity.astype(str) + ')'\n",
    "\n",
    "'''save file'''\n",
    "df_iTOL_cg.to_csv(dir_iTOL+\"color_gradient_and_color_ranges_iTOL_table.csv\",index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Shape plot\n",
    "Visualizes the gene group hits of core, tailoring and cluster specific fenes. The size reflext the size reflects the hgihest protein identity score from the gene group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''from cblaster table we load in  '''\n",
    "df_iTOL_sp = pd.DataFrame()\n",
    "df_iTOL_sp['Tree node ID'] = df_cblaster['Organism']\n",
    "df_iTOL_sp['Tree node label'] = df_cblaster['Organism']\n",
    "\n",
    "'''Now we want to assign IDs to the previously defined gene groups and join annotations into these groups'''\n",
    "gene_groups = {str(i+1).zfill(3)+\"-\"+str(len(group)).zfill(2): group for i, group in enumerate(gene_group_list)}\n",
    "'''Add groups as column names to the iTOL df'''\n",
    "for column_name in list(gene_groups):\n",
    "    df_iTOL_sp[column_name] = ''\n",
    "\n",
    "for i, group in gene_groups.items():\n",
    "    df_iTOL_sp[i] = df_cblaster[list(group)].max(axis=1)\n",
    "df_iTOL_sp = df_iTOL_sp.fillna(0)\n",
    "\n",
    "df_iTOL_sp.to_csv(dir_iTOL+\"shape_plot_data_iTOL.csv\",index=False)\n",
    "\n",
    "'''upload of this data in itol requires special header that to my knowladge cannot be made in one table, therefore sperate file is created:'''\n",
    "df_iTOL_cg_header = pd.DataFrame(list(df_iTOL_sp.columns[2:]), columns=['Field labels'])\n",
    "df_iTOL_cg_header[['Gene group', 'number of genes']] = df_iTOL_cg_header['Field labels'].str.split(\"-\", expand = True)\n",
    "df_iTOL_cg_header['number of genes'] = df_iTOL_cg_header['number of genes'].astype('int')\n",
    "'''color coding is done below based on nr of genes in each group'''\n",
    "hex_codes = []\n",
    "for gene_number in df_iTOL_cg_header['number of genes']:\n",
    "    color_strength = gene_number/max(df_iTOL_cg_header['number of genes'])\n",
    "    R_color = 255 - int(204*color_strength)\n",
    "    G_color = 255 - int(204*color_strength)\n",
    "    B_color = 255 - int(204*color_strength)\n",
    "    color_code = '#' + \"\".join(hex(val).split(\"0x\")[1].zfill(2) for val in [R_color, G_color, B_color])\n",
    "    hex_codes.append(color_code)\n",
    "\n",
    "df_iTOL_cg_header['Field colors'] = hex_codes\n",
    "df_iTOL_cg_header=df_iTOL_cg_header.drop(columns=['Gene group', 'number of genes']).transpose()\n",
    "\n",
    "df_iTOL_cg_header.to_csv(dir_iTOL+\"shape_plot_header_iTOL.csv\",index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.7 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "26de051ba29f2982a8de78e945f0abaf191376122a1563185a90213a26c5da77"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
