{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Targeted Cluster Analysis using cBLASTER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import altair as alt\n",
    "import glob\n",
    "from Bio import Phylo\n",
    "from Bio import SeqIO\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define Project and folder layour"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Project Name\n",
    "project=\"NCBI_P8A2_20230216\"\n",
    "BGC_group = 'Azoxy'\n",
    "dir_applied_tools= project+\"/00_applied_tools/\"\n",
    "dir_clinker = dir_applied_tools+\"clinker/\"\n",
    "dir_cblaster = dir_applied_tools+\"cblaster/\"\n",
    "dir_iTOL = dir_applied_tools+\"iTOL/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Make folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "directories = (dir_applied_tools, dir_clinker, dir_cblaster, dir_iTOL)\n",
    "# Check whether the specified path exists or not\n",
    "for path in directories:\n",
    "   isExist = os.path.exists(path)\n",
    "   if not isExist:\n",
    "      # Create a new directory because it does not exist\n",
    "      os.makedirs(path)\n",
    "      print(\"Directory is created - \"+path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading in cBLASTER and cLINKER results and creating overview table"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Code below was automatically taking all cBLASTER hits, Ideally we need to have core genes identified that i have done for now manually, meaning genes present in all of the input clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'EHGBBDDM_28800', 'PBHDNOKL_17950', 'PU648_09280', 'ady7', 'vlmA'},\n",
       " {'EHGBBDDM_28930', 'PBHDNOKL_17790', 'PU648_09235', 'ady3', 'vlmH'},\n",
       " {'EHGBBDDM_28900', 'PBHDNOKL_17820', 'PU648_09285', 'ady6', 'vlmO'},\n",
       " {'EHGBBDDM_28880', 'PBHDNOKL_17840', 'PU648_09275', 'ady8', 'vlmB'},\n",
       " {'EHGBBDDM_28890', 'PBHDNOKL_17830', 'PU648_09255', 'ady10', 'vlmR'},\n",
       " {'EHGBBDDM_28920', 'PBHDNOKL_17800', 'PU648_09245', 'ady4'},\n",
       " {'EHGBBDDM_28910', 'PBHDNOKL_17810', 'PU648_09250', 'ady5'},\n",
       " {'EHGBBDDM_28840', 'PBHDNOKL_17930', 'PU648_09270', 'ady9'},\n",
       " {'EHGBBDDM_28830', 'PBHDNOKL_17960', 'vlmI'},\n",
       " {'EHGBBDDM_28940', 'PBHDNOKL_17780', 'ady2'},\n",
       " {'EHGBBDDM_28970', 'PBHDNOKL_17730', 'orf+1'},\n",
       " {'EHGBBDDM_28960', 'PBHDNOKL_17700', 'PBHDNOKL_17710'},\n",
       " {'PBHDNOKL_17880', 'vlmC'},\n",
       " {'PBHDNOKL_17680', 'vlmJ'},\n",
       " {'PBHDNOKL_17660', 'vlmK'},\n",
       " {'PBHDNOKL_17890', 'vlmL'},\n",
       " {'PU648_09290', 'ady1'},\n",
       " {'PU648_09300', 'ady11'},\n",
       " {'PU648_09295', 'ady12'},\n",
       " {'EHGBBDDM_28980', 'PBHDNOKL_17750'},\n",
       " {'EHGBBDDM_28990', 'PBHDNOKL_17760'},\n",
       " {'EHGBBDDM_29000', 'PBHDNOKL_17770'},\n",
       " {'EHGBBDDM_28870', 'PBHDNOKL_17850'},\n",
       " {'EHGBBDDM_28860', 'PBHDNOKL_17860'},\n",
       " {'EHGBBDDM_28850', 'PBHDNOKL_17870'},\n",
       " {'EHGBBDDM_28810', 'PBHDNOKL_17940'},\n",
       " {'PBHDNOKL_17690', 'PU648_09370'}]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Code written together with thom to create the gene matrix automatically\n",
    "df_genes_list = pd.read_csv(dir_clinker +'alignments.csv',header=None,usecols=[0,1]).dropna()\n",
    "df_genes_list.rename( columns={0 :'Pair_gene_1'}, inplace=True )\n",
    "df_genes_list.rename( columns={1 :'Pair_gene_2'}, inplace=True )\n",
    "# Create an empty list\n",
    "genes_list =[]\n",
    "# Iterate over each row\n",
    "for rows in df_genes_list.itertuples():\n",
    "    # Create list for the current row\n",
    "    each_pair =[rows.Pair_gene_1, rows.Pair_gene_2]\n",
    "    # append the list to the final list\n",
    "    genes_list.append(each_pair)\n",
    "\n",
    "gene_group_list = [{genes_list[0][0]}]\n",
    "for genes in genes_list:\n",
    "    genes_added = False\n",
    "    for gene_group in gene_group_list:\n",
    "        if genes[0] in gene_group or genes[1] in gene_group:\n",
    "            gene_group.add(genes[1])\n",
    "            gene_group.add(genes[0])\n",
    "            genes_added = True\n",
    "    if genes_added == False:\n",
    "        gene_group_list.extend([{genes[0]}])\n",
    "        for gene_group in gene_group_list:\n",
    "            if genes[0] in gene_group or genes[1] in gene_group:\n",
    "                gene_group.add(genes[1])\n",
    "                gene_group.add(genes[0])\n",
    "sorted(gene_group_list)\n",
    "gene_group_list.sort(key = len, reverse=True)\n",
    "\n",
    "clean_gene_group_dictionary = {}\n",
    "for index, group in enumerate(gene_group_list, start = 1):\n",
    "    x = str(index).zfill(3) + \"-\" + str(len(group)).zfill(2)\n",
    "    temp_group_dict = dict.fromkeys(group, x)\n",
    "    clean_gene_group_dictionary.update(temp_group_dict)\n",
    "\n",
    "gene_group_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\matmal\\AppData\\Local\\Temp\\ipykernel_5784\\2726100085.py:6: FutureWarning: DataFrame.set_axis 'inplace' keyword is deprecated and will be removed in a future version. Use `obj = obj.set_axis(..., copy=False)` instead\n",
      "  clinker_hit_gene_df =pd.DataFrame(clinker_hit_gene_df).set_axis(['clinker_hit_genes'], axis=1, inplace=False).drop_duplicates()\n",
      "C:\\Users\\matmal\\AppData\\Local\\Temp\\ipykernel_5784\\2726100085.py:12: FutureWarning: DataFrame.set_axis 'inplace' keyword is deprecated and will be removed in a future version. Use `obj = obj.set_axis(..., copy=False)` instead\n",
      "  core_gene_df =pd.DataFrame(core_gene_df).set_axis(['core_genes'], axis=1, inplace=False)\n"
     ]
    }
   ],
   "source": [
    "# following code takes creates list of genes that share similartiy between the input clusters by analysing clinker alignment .csv file\n",
    "#      note, that if two of the same clusters are uploaded, all genes in the BGC will be marked as shared genes and therefore will not be unique.\n",
    "#           It could be considered to put score for each gene, eg. between how many clusters this gene is shared, thus indicating if it is essential or tailoring gene\n",
    "df = pd.read_csv(dir_clinker +'alignments.csv',header=None,usecols=[0,1])\n",
    "clinker_hit_gene_df = pd.concat([df.iloc[:,0],df.iloc[:,1]])\n",
    "clinker_hit_gene_df =pd.DataFrame(clinker_hit_gene_df).set_axis(['clinker_hit_genes'], axis=1, inplace=False).drop_duplicates()\n",
    "clinker_hit_gene_list = clinker_hit_gene_df['clinker_hit_genes'].values.tolist()\n",
    "clinker_hit_gene_df.to_csv(dir_clinker+\"clinker_hit_gene_df.csv\",index=False)\n",
    "\n",
    "#Core genes, table made manually\n",
    "core_gene_df = pd.read_csv(dir_clinker +'manual_Core.csv',header=None)\n",
    "core_gene_df =pd.DataFrame(core_gene_df).set_axis(['core_genes'], axis=1, inplace=False)\n",
    "core_gene_list = core_gene_df['core_genes'].values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cblaster processed files\n",
    "cblaster_sample_overview = pd.read_csv(dir_cblaster+'jobs.txt', header=None)\n",
    "cblaster_sample_overview.columns = ['Inputs']\n",
    "cblaster_inputs_list = list(cblaster_sample_overview['Inputs'])\n",
    "cblaster_inputs_list\n",
    "\n",
    "\n",
    "# Code below can be used but needs to be activated below.\n",
    "#Define minimum number of core genes identified by clinker for cluster to be kept\n",
    "minimum_core_genes = 4\n",
    "minimum_clinker_gene_hits = 5\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\matmal\\AppData\\Local\\Temp\\ipykernel_5784\\2416689173.py:14: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.\n",
      "  df.drop(df.columns.difference(clinker_hit_gene_list), 1, inplace=True)\n",
      "C:\\Users\\matmal\\AppData\\Roaming\\Python\\Python310\\site-packages\\pandas\\core\\apply.py:884: RuntimeWarning: Mean of empty slice\n",
      "  results[i] = self.f(v)\n",
      "C:\\Users\\matmal\\AppData\\Local\\Temp\\ipykernel_5784\\2416689173.py:30: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.\n",
      "  df.drop(df.columns.difference(core_gene_list), 1, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>cblaster_input</th>\n",
       "      <th>gene_hits</th>\n",
       "      <th>clinker_gene_hits</th>\n",
       "      <th>%_clinker_gene_identity</th>\n",
       "      <th>%_average_gene_identity</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Scaffold</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>...</th>\n",
       "      <th>PU648_09325</th>\n",
       "      <th>PU648_09330</th>\n",
       "      <th>PU648_09335</th>\n",
       "      <th>PU648_09340</th>\n",
       "      <th>PU648_09345</th>\n",
       "      <th>PU648_09350</th>\n",
       "      <th>PU648_09355</th>\n",
       "      <th>PU648_09360</th>\n",
       "      <th>PU648_09365</th>\n",
       "      <th>PU648_09370</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1832</td>\n",
       "      <td>P8-A2_Azodyrecin</td>\n",
       "      <td>36</td>\n",
       "      <td>12</td>\n",
       "      <td>99.908333</td>\n",
       "      <td>99.930556</td>\n",
       "      <td>GCF_007113545.2</td>\n",
       "      <td>NZ_CP041602.2</td>\n",
       "      <td>9821626</td>\n",
       "      <td>9898584</td>\n",
       "      <td>...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1825</td>\n",
       "      <td>P8-A2_Azodyrecin</td>\n",
       "      <td>36</td>\n",
       "      <td>12</td>\n",
       "      <td>99.908333</td>\n",
       "      <td>99.930556</td>\n",
       "      <td>GCF_007113485.2</td>\n",
       "      <td>NZ_CP041613.2</td>\n",
       "      <td>10165091</td>\n",
       "      <td>10242028</td>\n",
       "      <td>...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1835</td>\n",
       "      <td>P8-A2_Azodyrecin</td>\n",
       "      <td>36</td>\n",
       "      <td>12</td>\n",
       "      <td>99.675000</td>\n",
       "      <td>99.694444</td>\n",
       "      <td>GCF_007113565.2</td>\n",
       "      <td>NZ_CP041604.2</td>\n",
       "      <td>9832598</td>\n",
       "      <td>9910604</td>\n",
       "      <td>...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>99.9</td>\n",
       "      <td>100.0</td>\n",
       "      <td>99.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>99.8</td>\n",
       "      <td>99.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1837</td>\n",
       "      <td>P8-A2_Azodyrecin</td>\n",
       "      <td>36</td>\n",
       "      <td>12</td>\n",
       "      <td>99.583333</td>\n",
       "      <td>99.694444</td>\n",
       "      <td>GCF_007113585.2</td>\n",
       "      <td>NZ_CP041609.2</td>\n",
       "      <td>9576908</td>\n",
       "      <td>9652128</td>\n",
       "      <td>...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>99.9</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>99.8</td>\n",
       "      <td>99.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4591</td>\n",
       "      <td>P8-A2_Azodyrecin</td>\n",
       "      <td>34</td>\n",
       "      <td>12</td>\n",
       "      <td>98.141667</td>\n",
       "      <td>92.144444</td>\n",
       "      <td>GCF_900112965.1</td>\n",
       "      <td>NZ_FONR01000012</td>\n",
       "      <td>138273</td>\n",
       "      <td>176623</td>\n",
       "      <td>...</td>\n",
       "      <td>90.5</td>\n",
       "      <td>98.6</td>\n",
       "      <td>99.8</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>97.8</td>\n",
       "      <td>99.2</td>\n",
       "      <td>97.9</td>\n",
       "      <td>98.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>238</td>\n",
       "      <td>LC712332_Azodyrecin</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>51.425000</td>\n",
       "      <td>31.646154</td>\n",
       "      <td>GCF_014649335.1</td>\n",
       "      <td>NZ_BMSV01000002</td>\n",
       "      <td>727788</td>\n",
       "      <td>744654</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>82</td>\n",
       "      <td>LC712332_Azodyrecin</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>49.433333</td>\n",
       "      <td>22.815385</td>\n",
       "      <td>GCF_002214185.1</td>\n",
       "      <td>NZ_CP022161.1</td>\n",
       "      <td>5094814</td>\n",
       "      <td>5118039</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>1007</td>\n",
       "      <td>AY116644_valanimycin</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>77.710000</td>\n",
       "      <td>66.500000</td>\n",
       "      <td>GCF_013414305.1</td>\n",
       "      <td>NZ_CP058693.1</td>\n",
       "      <td>118309</td>\n",
       "      <td>138736</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>1579</td>\n",
       "      <td>AY116644_valanimycin</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>80.611111</td>\n",
       "      <td>52.815000</td>\n",
       "      <td>GCF_022698285.1</td>\n",
       "      <td>c00001_NZ_JALD..</td>\n",
       "      <td>9251988</td>\n",
       "      <td>9266505</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>1083</td>\n",
       "      <td>AY116644_valanimycin</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>70.133333</td>\n",
       "      <td>42.410000</td>\n",
       "      <td>GCF_014648775.1</td>\n",
       "      <td>NZ_BMRT01000013</td>\n",
       "      <td>62548</td>\n",
       "      <td>77960</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>352 rows × 180 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index        cblaster_input gene_hits  clinker_gene_hits  \\\n",
       "0     1832      P8-A2_Azodyrecin        36                 12   \n",
       "1     1825      P8-A2_Azodyrecin        36                 12   \n",
       "2     1835      P8-A2_Azodyrecin        36                 12   \n",
       "3     1837      P8-A2_Azodyrecin        36                 12   \n",
       "4     4591      P8-A2_Azodyrecin        34                 12   \n",
       "..     ...                   ...       ...                ...   \n",
       "427    238   LC712332_Azodyrecin         8                  8   \n",
       "428     82   LC712332_Azodyrecin         6                  6   \n",
       "429   1007  AY116644_valanimycin        17                 10   \n",
       "430   1579  AY116644_valanimycin        13                  9   \n",
       "431   1083  AY116644_valanimycin        12                  9   \n",
       "\n",
       "     %_clinker_gene_identity  %_average_gene_identity         Organism  \\\n",
       "0                  99.908333                99.930556  GCF_007113545.2   \n",
       "1                  99.908333                99.930556  GCF_007113485.2   \n",
       "2                  99.675000                99.694444  GCF_007113565.2   \n",
       "3                  99.583333                99.694444  GCF_007113585.2   \n",
       "4                  98.141667                92.144444  GCF_900112965.1   \n",
       "..                       ...                      ...              ...   \n",
       "427                51.425000                31.646154  GCF_014649335.1   \n",
       "428                49.433333                22.815385  GCF_002214185.1   \n",
       "429                77.710000                66.500000  GCF_013414305.1   \n",
       "430                80.611111                52.815000  GCF_022698285.1   \n",
       "431                70.133333                42.410000  GCF_014648775.1   \n",
       "\n",
       "             Scaffold     Start       End  ...  PU648_09325  PU648_09330  \\\n",
       "0       NZ_CP041602.2   9821626   9898584  ...        100.0        100.0   \n",
       "1       NZ_CP041613.2  10165091  10242028  ...        100.0        100.0   \n",
       "2       NZ_CP041604.2   9832598   9910604  ...        100.0         99.9   \n",
       "3       NZ_CP041609.2   9576908   9652128  ...        100.0         99.9   \n",
       "4     NZ_FONR01000012    138273    176623  ...         90.5         98.6   \n",
       "..                ...       ...       ...  ...          ...          ...   \n",
       "427   NZ_BMSV01000002    727788    744654  ...          NaN          NaN   \n",
       "428     NZ_CP022161.1   5094814   5118039  ...          NaN          NaN   \n",
       "429     NZ_CP058693.1    118309    138736  ...          NaN          NaN   \n",
       "430  c00001_NZ_JALD..   9251988   9266505  ...          NaN          NaN   \n",
       "431   NZ_BMRT01000013     62548     77960  ...          NaN          NaN   \n",
       "\n",
       "     PU648_09335  PU648_09340  PU648_09345  PU648_09350  PU648_09355  \\\n",
       "0          100.0        100.0        100.0        100.0        100.0   \n",
       "1          100.0        100.0        100.0        100.0        100.0   \n",
       "2          100.0         99.5        100.0        100.0        100.0   \n",
       "3          100.0        100.0        100.0        100.0        100.0   \n",
       "4           99.8        100.0        100.0        100.0         97.8   \n",
       "..           ...          ...          ...          ...          ...   \n",
       "427          NaN          NaN          NaN          NaN          NaN   \n",
       "428          NaN          NaN          NaN          NaN          NaN   \n",
       "429          NaN          NaN          NaN          NaN          NaN   \n",
       "430          NaN          NaN          NaN          NaN          NaN   \n",
       "431          NaN          NaN          NaN          NaN          NaN   \n",
       "\n",
       "     PU648_09360  PU648_09365  PU648_09370  \n",
       "0          100.0        100.0        100.0  \n",
       "1          100.0        100.0        100.0  \n",
       "2          100.0         99.8         99.3  \n",
       "3          100.0         99.8         99.5  \n",
       "4           99.2         97.9         98.3  \n",
       "..           ...          ...          ...  \n",
       "427          NaN          NaN          NaN  \n",
       "428          NaN          NaN          NaN  \n",
       "429          NaN          NaN          NaN  \n",
       "430          NaN          NaN          NaN  \n",
       "431          NaN          NaN          NaN  \n",
       "\n",
       "[352 rows x 180 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cblaster output file\n",
    "cblaster_output = \"data.csv\"\n",
    "# output tables\n",
    "df_cblaster = pd.DataFrame(columns=['cblaster_input','gene_hits','clinker_gene_hits','%_clinker_gene_identity','%_average_gene_identity','Organism','Scaffold','Start','End','Score'])\n",
    "for cblaster_input in cblaster_inputs_list:\n",
    "    df=pd.read_csv(dir_cblaster+cblaster_input+\"/\"+cblaster_output)\n",
    "    #df['%_cluster_gene_hits']=df.iloc[:, 5:].apply(np.count_nonzero, axis=1)/len(df.iloc[:, 5:].columns)\n",
    "    df['gene_hits']=df.iloc[:, 5:].apply(np.count_nonzero, axis=1)\n",
    "    df['%_average_gene_identity']=df.iloc[:, 5:-1].apply(np.nanmean, axis=1)\n",
    "    df['cblaster_input'] = cblaster_input\n",
    "    df_cblaster= pd.concat([df_cblaster, df])\n",
    "# make subtable of cblaster outputs, filtering only gene hits by clinker\n",
    "df = df_cblaster.copy()\n",
    "df.drop(df.columns.difference(clinker_hit_gene_list), 1, inplace=True)\n",
    "\n",
    "\n",
    "df['clinker_gene_hits'] = df[df > 0].count(axis=1)\n",
    "df['%_average_gene_identity'] = df[df > 0].iloc[:,:-1].apply(np.nanmean, axis=1)\n",
    "df_cblaster['clinker_gene_hits'] = df['clinker_gene_hits']\n",
    "df_cblaster['%_clinker_gene_identity'] = df['%_average_gene_identity']\n",
    "# sort the table and remove dublicated genome hits for different clusters, keeping only one with most hits\n",
    "df_cblaster = df_cblaster.sort_values(by ='clinker_gene_hits', ascending = False)\n",
    "df_cblaster = df_cblaster.drop_duplicates(subset=['Organism','Scaffold'], keep='first')\n",
    "df_cblaster = df_cblaster[df_cblaster['clinker_gene_hits'] >= minimum_clinker_gene_hits]\n",
    "df_cblaster = df_cblaster.sort_values(by =['cblaster_input','clinker_gene_hits'], ascending = False)\n",
    "df_cblaster = df_cblaster.reset_index()\n",
    "\n",
    "# Drop rows that do not contain core genes that are shered between all cblaster outputs\n",
    "df = df_cblaster.copy()\n",
    "df.drop(df.columns.difference(core_gene_list), 1, inplace=True)\n",
    "df=df.replace(0, np.nan)\n",
    "df.dropna(axis = 0, how = 'all', inplace = True)\n",
    "df['core_gene_hits'] = df[df > 0].count(axis=1)\n",
    "df = df[df['core_gene_hits'] >= minimum_core_genes]\n",
    "df = pd.DataFrame(index=df.index)\n",
    "df_cblaster = pd.merge(df, df_cblaster, left_index=True, right_index=True, how=\"inner\")\n",
    "\n",
    "# Save table\n",
    "df_cblaster.to_csv(dir_cblaster+\"df_cblaster.csv\")\n",
    "df_cblaster"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iTOL visualization and splitting data into sub-tables\n",
    "- To visalize the data, create a table where Each cluster is given its own color\n",
    "    - Group the most similar clusters with XX cutoff to be responsible for production of best mach cBLASTER cluster\n",
    "        - Probably best to use max cluster hits / with core gene hits to get estimated ratio.\n",
    "            - This can create problems if input cluster is not part of the diamond database and better would be to use clinker identifited gene coutn. Requires to make table with cblaster_input and clinker hits\n",
    "    - Low similarity clusters are offcolored to be something similar to target input\n",
    "    - very-low similarity clusters are graphed as other related clusters."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## color gradient\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Manually create label annotation names and rbg color codes that are linked to cblaster input names"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test with Thom\n",
    "#Define cluster colors\n",
    "df_labels = cblaster_sample_overview.copy()\n",
    "df_labels['Label'] = ['BGC0001066','BGC0001819','rsn']\n",
    "df_labels['Color code for rbg'] = ['0,158,115','230,159,0','0,114,178']\n",
    "df_labels['HEX'] = ['#009e73','#e69f00','#0072b2']\n",
    "df_labels['nr_of_core_genes'] = [10,7,10]\n",
    "df_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inputs</th>\n",
       "      <th>Label</th>\n",
       "      <th>Color code for rbg</th>\n",
       "      <th>HEX</th>\n",
       "      <th>nr_of_core_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AY116644_valanimycin</td>\n",
       "      <td>Valanimycin</td>\n",
       "      <td>0,158,115</td>\n",
       "      <td>#009e73</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LC712332_Azodyrecin</td>\n",
       "      <td>Azodyrecin A-F</td>\n",
       "      <td>230,159,0</td>\n",
       "      <td>#e69f00</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NZ_AP018517.1_KA57-A</td>\n",
       "      <td>KA57-A</td>\n",
       "      <td>0,114,178</td>\n",
       "      <td>#0072b2</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NZ_CP027306.1_Elaiomycin</td>\n",
       "      <td>Eliomycin</td>\n",
       "      <td>86,180,233</td>\n",
       "      <td>#56b4e9</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P8-A2_Azodyrecin</td>\n",
       "      <td>Azodyrecin A-C</td>\n",
       "      <td>213,94,0</td>\n",
       "      <td>#d55e00</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Inputs           Label Color code for rbg      HEX  \\\n",
       "0      AY116644_valanimycin     Valanimycin          0,158,115  #009e73   \n",
       "1       LC712332_Azodyrecin  Azodyrecin A-F          230,159,0  #e69f00   \n",
       "2      NZ_AP018517.1_KA57-A          KA57-A          0,114,178  #0072b2   \n",
       "3  NZ_CP027306.1_Elaiomycin       Eliomycin         86,180,233  #56b4e9   \n",
       "4          P8-A2_Azodyrecin  Azodyrecin A-C           213,94,0  #d55e00   \n",
       "\n",
       "   nr_of_core_genes  \n",
       "0                10  \n",
       "1                13  \n",
       "2                25  \n",
       "3                19  \n",
       "4                12  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Define cluster colors\n",
    "df_labels = cblaster_sample_overview.copy()\n",
    "df_labels['Label'] = ['Valanimycin','Azodyrecin A-F','KA57-A','Eliomycin','Azodyrecin A-C']\n",
    "df_labels['Color code for rbg'] = ['0,158,115','230,159,0','0,114,178','86,180,233','213,94,0']\n",
    "df_labels['HEX'] = ['#009e73','#e69f00','#0072b2','#56b4e9','#d55e00']\n",
    "df_labels['nr_of_core_genes'] = [10,13,25,19,12]\n",
    "df_labels\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create iTOL annotation table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tree node (bgcflow input file name)</th>\n",
       "      <th>Color (rgba(255, 255, 255, 1.000))</th>\n",
       "      <th>gene_hits</th>\n",
       "      <th>clinker_gene_hits</th>\n",
       "      <th>%_clinker_gene_identity</th>\n",
       "      <th>%_average_gene_identity</th>\n",
       "      <th>Label</th>\n",
       "      <th>Color code for rbg</th>\n",
       "      <th>HEX</th>\n",
       "      <th>nr_of_core_genes</th>\n",
       "      <th>%_clinker_gene_hits</th>\n",
       "      <th>transperency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AY116644_valanimycin</th>\n",
       "      <td>GCF_013414305.1</td>\n",
       "      <td>rgba(0,158,115,0.7771000000000001)</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>77.710000</td>\n",
       "      <td>66.500000</td>\n",
       "      <td>Valanimycin</td>\n",
       "      <td>0,158,115</td>\n",
       "      <td>#009e73</td>\n",
       "      <td>10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.777100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AY116644_valanimycin</th>\n",
       "      <td>GCF_022698285.1</td>\n",
       "      <td>rgba(0,158,115,0.7255000000000001)</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>80.611111</td>\n",
       "      <td>52.815000</td>\n",
       "      <td>Valanimycin</td>\n",
       "      <td>0,158,115</td>\n",
       "      <td>#009e73</td>\n",
       "      <td>10</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.725500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AY116644_valanimycin</th>\n",
       "      <td>GCF_014648775.1</td>\n",
       "      <td>rgba(0,158,115,0.6312000000000001)</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>70.133333</td>\n",
       "      <td>42.410000</td>\n",
       "      <td>Valanimycin</td>\n",
       "      <td>0,158,115</td>\n",
       "      <td>#009e73</td>\n",
       "      <td>10</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.631200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LC712332_Azodyrecin</th>\n",
       "      <td>GCF_019933235.1</td>\n",
       "      <td>rgba(230,159,0,0.9677692307692307)</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>96.776923</td>\n",
       "      <td>96.776923</td>\n",
       "      <td>Azodyrecin A-F</td>\n",
       "      <td>230,159,0</td>\n",
       "      <td>#e69f00</td>\n",
       "      <td>13</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.967769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LC712332_Azodyrecin</th>\n",
       "      <td>GCF_009176265.1</td>\n",
       "      <td>rgba(230,159,0,0.9677692307692307)</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>96.776923</td>\n",
       "      <td>96.776923</td>\n",
       "      <td>Azodyrecin A-F</td>\n",
       "      <td>230,159,0</td>\n",
       "      <td>#e69f00</td>\n",
       "      <td>13</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.967769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8-A2_Azodyrecin</th>\n",
       "      <td>GCF_026342395.1</td>\n",
       "      <td>rgba(0, 0, 0,0.4227499999999999)</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>63.412500</td>\n",
       "      <td>18.027778</td>\n",
       "      <td>Azodyrecin A-C</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>#d55e00</td>\n",
       "      <td>12</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.422750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8-A2_Azodyrecin</th>\n",
       "      <td>GCF_025398775.1</td>\n",
       "      <td>rgba(0, 0, 0,0.4004166666666666)</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>60.062500</td>\n",
       "      <td>18.377778</td>\n",
       "      <td>Azodyrecin A-C</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>#d55e00</td>\n",
       "      <td>12</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.400417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8-A2_Azodyrecin</th>\n",
       "      <td>GCF_027947595.1</td>\n",
       "      <td>rgba(0, 0, 0,0.4174166666666666)</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>62.612500</td>\n",
       "      <td>18.883333</td>\n",
       "      <td>Azodyrecin A-C</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>#d55e00</td>\n",
       "      <td>12</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.417417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8-A2_Azodyrecin</th>\n",
       "      <td>GCF_001514285.1</td>\n",
       "      <td>rgba(0, 0, 0,0.2873333333333334)</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>49.257143</td>\n",
       "      <td>10.497222</td>\n",
       "      <td>Azodyrecin A-C</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>#d55e00</td>\n",
       "      <td>12</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>0.287333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8-A2_Azodyrecin</th>\n",
       "      <td>GCF_001514205.1</td>\n",
       "      <td>rgba(0, 0, 0,0.25016666666666665)</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>50.033333</td>\n",
       "      <td>9.302778</td>\n",
       "      <td>Azodyrecin A-C</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>#d55e00</td>\n",
       "      <td>12</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.250167</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>352 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Tree node (bgcflow input file name)  \\\n",
       "AY116644_valanimycin                     GCF_013414305.1   \n",
       "AY116644_valanimycin                     GCF_022698285.1   \n",
       "AY116644_valanimycin                     GCF_014648775.1   \n",
       "LC712332_Azodyrecin                      GCF_019933235.1   \n",
       "LC712332_Azodyrecin                      GCF_009176265.1   \n",
       "...                                                  ...   \n",
       "P8-A2_Azodyrecin                         GCF_026342395.1   \n",
       "P8-A2_Azodyrecin                         GCF_025398775.1   \n",
       "P8-A2_Azodyrecin                         GCF_027947595.1   \n",
       "P8-A2_Azodyrecin                         GCF_001514285.1   \n",
       "P8-A2_Azodyrecin                         GCF_001514205.1   \n",
       "\n",
       "                      Color (rgba(255, 255, 255, 1.000)) gene_hits  \\\n",
       "AY116644_valanimycin  rgba(0,158,115,0.7771000000000001)        17   \n",
       "AY116644_valanimycin  rgba(0,158,115,0.7255000000000001)        13   \n",
       "AY116644_valanimycin  rgba(0,158,115,0.6312000000000001)        12   \n",
       "LC712332_Azodyrecin   rgba(230,159,0,0.9677692307692307)        13   \n",
       "LC712332_Azodyrecin   rgba(230,159,0,0.9677692307692307)        13   \n",
       "...                                                  ...       ...   \n",
       "P8-A2_Azodyrecin        rgba(0, 0, 0,0.4227499999999999)        10   \n",
       "P8-A2_Azodyrecin        rgba(0, 0, 0,0.4004166666666666)        11   \n",
       "P8-A2_Azodyrecin        rgba(0, 0, 0,0.4174166666666666)        11   \n",
       "P8-A2_Azodyrecin        rgba(0, 0, 0,0.2873333333333334)         8   \n",
       "P8-A2_Azodyrecin       rgba(0, 0, 0,0.25016666666666665)         7   \n",
       "\n",
       "                      clinker_gene_hits  %_clinker_gene_identity  \\\n",
       "AY116644_valanimycin                 10                77.710000   \n",
       "AY116644_valanimycin                  9                80.611111   \n",
       "AY116644_valanimycin                  9                70.133333   \n",
       "LC712332_Azodyrecin                  13                96.776923   \n",
       "LC712332_Azodyrecin                  13                96.776923   \n",
       "...                                 ...                      ...   \n",
       "P8-A2_Azodyrecin                      8                63.412500   \n",
       "P8-A2_Azodyrecin                      8                60.062500   \n",
       "P8-A2_Azodyrecin                      8                62.612500   \n",
       "P8-A2_Azodyrecin                      7                49.257143   \n",
       "P8-A2_Azodyrecin                      6                50.033333   \n",
       "\n",
       "                      %_average_gene_identity           Label  \\\n",
       "AY116644_valanimycin                66.500000     Valanimycin   \n",
       "AY116644_valanimycin                52.815000     Valanimycin   \n",
       "AY116644_valanimycin                42.410000     Valanimycin   \n",
       "LC712332_Azodyrecin                 96.776923  Azodyrecin A-F   \n",
       "LC712332_Azodyrecin                 96.776923  Azodyrecin A-F   \n",
       "...                                       ...             ...   \n",
       "P8-A2_Azodyrecin                    18.027778  Azodyrecin A-C   \n",
       "P8-A2_Azodyrecin                    18.377778  Azodyrecin A-C   \n",
       "P8-A2_Azodyrecin                    18.883333  Azodyrecin A-C   \n",
       "P8-A2_Azodyrecin                    10.497222  Azodyrecin A-C   \n",
       "P8-A2_Azodyrecin                     9.302778  Azodyrecin A-C   \n",
       "\n",
       "                     Color code for rbg      HEX  nr_of_core_genes  \\\n",
       "AY116644_valanimycin          0,158,115  #009e73                10   \n",
       "AY116644_valanimycin          0,158,115  #009e73                10   \n",
       "AY116644_valanimycin          0,158,115  #009e73                10   \n",
       "LC712332_Azodyrecin           230,159,0  #e69f00                13   \n",
       "LC712332_Azodyrecin           230,159,0  #e69f00                13   \n",
       "...                                 ...      ...               ...   \n",
       "P8-A2_Azodyrecin                0, 0, 0  #d55e00                12   \n",
       "P8-A2_Azodyrecin                0, 0, 0  #d55e00                12   \n",
       "P8-A2_Azodyrecin                0, 0, 0  #d55e00                12   \n",
       "P8-A2_Azodyrecin                0, 0, 0  #d55e00                12   \n",
       "P8-A2_Azodyrecin                0, 0, 0  #d55e00                12   \n",
       "\n",
       "                      %_clinker_gene_hits  transperency  \n",
       "AY116644_valanimycin             1.000000      0.777100  \n",
       "AY116644_valanimycin             0.900000      0.725500  \n",
       "AY116644_valanimycin             0.900000      0.631200  \n",
       "LC712332_Azodyrecin              1.000000      0.967769  \n",
       "LC712332_Azodyrecin              1.000000      0.967769  \n",
       "...                                   ...           ...  \n",
       "P8-A2_Azodyrecin                 0.666667      0.422750  \n",
       "P8-A2_Azodyrecin                 0.666667      0.400417  \n",
       "P8-A2_Azodyrecin                 0.666667      0.417417  \n",
       "P8-A2_Azodyrecin                 0.583333      0.287333  \n",
       "P8-A2_Azodyrecin                 0.500000      0.250167  \n",
       "\n",
       "[352 rows x 12 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Following code is used to split the data in the [df_cblaster] based on specific criteria\n",
    "\n",
    "    # Criteria (currently not used)\n",
    "bgc_product_core_genes_of_clinker = 0.95\n",
    "bgc_prodyct_core_gene_identity = 0.9\n",
    "\n",
    "bgc_cloesly_related_core_genes_of_clinker = 0.85\n",
    "gc_cloesly_related_core_gene_identity = 0.7\n",
    "\n",
    "bgc_other_genes_of_clinker = 0.1\n",
    "\n",
    "    # Create three DataFrames\n",
    "\n",
    "df_iTOL_cg = pd.DataFrame(columns=['Tree node (bgcflow input file name)','Color (rgba(255, 255, 255, 1.000))','Label'])\n",
    "#Populate the table with df_cblaster results\n",
    "#\n",
    "\n",
    "df_iTOL_cg['Tree node (bgcflow input file name)'] = df_cblaster['Organism'] \n",
    "df_iTOL_cg['Label'] = df_cblaster['cblaster_input']\n",
    "df_iTOL_cg['gene_hits'] = df_cblaster['gene_hits']\n",
    "df_iTOL_cg['clinker_gene_hits'] = df_cblaster['clinker_gene_hits']\n",
    "df_iTOL_cg['%_clinker_gene_identity'] = df_cblaster['%_clinker_gene_identity']\n",
    "df_iTOL_cg['%_average_gene_identity'] = df_cblaster['%_average_gene_identity']\n",
    "\n",
    "#Mrege in manually made Label translation table, with rbg codes, nr of core genes and label ids. Table index is changed to link the two tables\n",
    "#\n",
    "df_iTOL_cg = pd.merge(df_iTOL_cg.set_index('Label'),df_labels.set_index('Inputs'),left_index=True, right_index=True)\n",
    "df_iTOL_cg['%_clinker_gene_hits'] = df_iTOL_cg['clinker_gene_hits'] / df_iTOL_cg['nr_of_core_genes']\n",
    "#Transperency defines the number of core genes hit\n",
    "df_iTOL_cg['transperency'] = df_iTOL_cg['%_clinker_gene_hits']*df_iTOL_cg['%_clinker_gene_identity']/100\n",
    "\n",
    "#We want to color the hits based on the similarity to the original cluster aswell as transperancy product that is defined as nr of core hits X core gene identity, thus estimating similarity between two clusters. If transperency is below XX % then we wnat to set color to black. \n",
    "similarity_score = 0.5\n",
    "\n",
    "df_iTOL_cg.loc[df_iTOL_cg['transperency'] <= similarity_score, 'Color code for rbg'] = '0, 0, 0'\n",
    "df_iTOL_cg['Color (rgba(255, 255, 255, 1.000))'] = 'rgba('+df_iTOL_cg['Color code for rbg']+','+df_iTOL_cg['transperency'].apply(str)+')'\n",
    "\n",
    "\n",
    "#Create the final table for iTOL export\n",
    "#\n",
    "df_iTOL_color_gradient = pd.DataFrame(columns=['Tree node','Color','Label'])\n",
    "df_iTOL_color_gradient['Tree node'] = df_iTOL_cg['Tree node (bgcflow input file name)']\n",
    "df_iTOL_color_gradient['Color'] = df_iTOL_cg['Color (rgba(255, 255, 255, 1.000))']\n",
    "df_iTOL_color_gradient['Label'] = df_iTOL_cg['Label']\n",
    "df_iTOL_color_gradient=df_iTOL_color_gradient.reset_index(drop=True)\n",
    "df_iTOL_color_gradient.to_csv(dir_iTOL+\"color_gradient.csv\",index=False)\n",
    "\n",
    "\n",
    "df_iTOL_cg"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Shape plot\n",
    "Use to visualize the gene architecture, eather by classifying which of the core shared genes belong to what function, thereby we can map percentage of core available\n",
    "\n",
    "note: To make shape plot, make DF that contains 1st column Genome ID and remaining columns each of the group to be visualized. Shapes and colors are set manually in iTOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Old code\n",
    "df_clinker_alignment_manual = pd.read_csv(dir_clinker + 'alignments_manual.csv')\n",
    "df_clinker_alignment_manual = df_clinker_alignment_manual.iloc[:,1:6].transpose().reset_index().drop('index', axis=1).transpose()\n",
    "df_clinker_alignment_manual['id'] = np.arange(len(df_clinker_alignment_manual)) // 1+ 1\n",
    "df_clinker_alignment_manual = df_clinker_alignment_manual.set_index('id').transpose()\n",
    "\n",
    "\n",
    "# create clean gene dictionary\n",
    "df_clinker_alignment_manual = df_clinker_alignment_manual.transpose().to_dict('tight')\n",
    "gene_groups = df_clinker_alignment_manual['data']\n",
    "clean_gene_group_list = []\n",
    "for group in gene_groups:\n",
    "    group = [x for x in group if str(x) != 'nan']\n",
    "    clean_gene_group_list.append(group)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code written together with thom to create the gene matrix automatically\n",
    "df_genes_list = pd.read_csv(dir_clinker +'alignments.csv',header=None,usecols=[0,1]).dropna()\n",
    "df_genes_list.rename( columns={0 :'Pair_gene_1'}, inplace=True )\n",
    "df_genes_list.rename( columns={1 :'Pair_gene_2'}, inplace=True )\n",
    "# Create an empty list\n",
    "genes_list =[]\n",
    "# Iterate over each row\n",
    "for rows in df_genes_list.itertuples():\n",
    "    # Create list for the current row\n",
    "    each_pair =[rows.Pair_gene_1, rows.Pair_gene_2]\n",
    "    # append the list to the final list\n",
    "    genes_list.append(each_pair)\n",
    "\n",
    "gene_group_list = [{genes_list[0][0]}]\n",
    "for genes in genes_list:\n",
    "    genes_added = False\n",
    "    for gene_group in gene_group_list:\n",
    "        if genes[0] in gene_group or genes[1] in gene_group:\n",
    "            gene_group.add(genes[1])\n",
    "            gene_group.add(genes[0])\n",
    "            genes_added = True\n",
    "    if genes_added == False:\n",
    "        gene_group_list.extend([{genes[0]}])\n",
    "        for gene_group in gene_group_list:\n",
    "            if genes[0] in gene_group or genes[1] in gene_group:\n",
    "                gene_group.add(genes[1])\n",
    "                gene_group.add(genes[0])\n",
    "sorted(gene_group_list)\n",
    "gene_group_list.sort(key = len, reverse=True)\n",
    "\n",
    "clean_gene_group_dictionary = {}\n",
    "for index, group in enumerate(gene_group_list, start = 1):\n",
    "    x = str(index).zfill(3) + \"-\" + str(len(group)).zfill(2)\n",
    "    temp_group_dict = dict.fromkeys(group, x)\n",
    "    clean_gene_group_dictionary.update(temp_group_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# loading in cBLASTER results into a list of gene hits to combine with clinker output\n",
    "df2_cblaster = df_cblaster.set_index('Organism')\n",
    "df2_cblaster = df2_cblaster.iloc[:,10:]\n",
    "    # DATA LOSS!!! REMOVAL OF DUBLICATES!!! Organism cant be used as index incase same cluster repeats 2x or more times in organism.\n",
    "df2_cblaster = df2_cblaster[~df2_cblaster.index.duplicated(keep='first')]\n",
    "df2_cblaster['Organism'] = df2_cblaster.index\n",
    "df2_cblaster\n",
    "df2_cblaster = pd.melt(df2_cblaster, id_vars=['Organism'],\n",
    "        var_name='Annotation', value_name='%_Identity')\n",
    "df2_cblaster = df2_cblaster.dropna(how='any')\n",
    "df2_cblaster = df2_cblaster[(df2_cblaster != 0).all(1)]\n",
    "\n",
    "\n",
    "df2_cblaster['gene group'] = df2_cblaster[\"Annotation\"].map(clean_gene_group_dictionary)\n",
    "df2_cblaster = df2_cblaster.dropna(how='any').drop(['Annotation'], axis=1)\n",
    "# error caused by one gene having two homologue copies and this appearing twice in the dictionary\n",
    "# keep only first indexed gene incase there are two genes in the organism belonging to the same cluster.\n",
    "df2_cblaster = df2_cblaster.pivot_table(index ='Organism', columns='gene group', values='%_Identity').fillna(0)\n",
    "df2_cblaster.to_csv(dir_iTOL+\"shape_plot.csv\")\n",
    "\n",
    "hex_df = pd.DataFrame(list(df2_cblaster.columns), columns=['Field labels'])\n",
    "hex_df[['Gene group', 'number of genes']] = hex_df['Field labels'].str.split(\"-\", expand = True)\n",
    "hex_df['number of genes'] = hex_df['number of genes'].astype('int')\n",
    "hex_codes = []\n",
    "for gene_number in hex_df['number of genes']:\n",
    "    color_strength = gene_number/max(hex_df['number of genes'])\n",
    "    R_color = 255 - int(255*color_strength)\n",
    "    G_color = 171\n",
    "    B_color = 210\n",
    "    color_code = '#' + \"\".join(hex(val).split(\"0x\")[1].zfill(2) for val in [R_color, G_color, B_color])\n",
    "    hex_codes.append(color_code)\n",
    "\n",
    "hex_df['Field colors'] = hex_codes\n",
    "hex_df=hex_df.drop(columns=['Gene group', 'number of genes']).transpose()\n",
    "\n",
    "hex_df.to_csv(dir_iTOL+\"headder.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame([['Dollar Sales', 'Unit Sales', 'Dollar Sales', 'Unit Sales'], \n",
    "                   [1, 2, 3, 4], [5, 6, 7, 8]], columns=[2016, 2016, 2015, 2015])\n",
    "\n",
    "new_labels = pd.MultiIndex.from_arrays([df.columns, df.iloc[0]], names=['Year', 'Sales'])\n",
    "df1 = df.set_axis(new_labels, axis=1).iloc[1:]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.7 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "26de051ba29f2982a8de78e945f0abaf191376122a1563185a90213a26c5da77"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
