# BGC_phylo_mapping

This package was made to map the BGC onto whole genome phylogenetic tree. The package is ment to be used as addon to **bgcflow** snakemake package (https://github.com/NBChub/bgcflow), where you can specify specific analysis to be done onto set of genomes in structured and systematic way. using **bgcflow** you have option to create phylogenetic tree based on input genomes aswell as cblaster database of input genomes. The **BGC_phly_mapping** package thereafter searches for specific target clusters and creates datatables for iTOL visualization.

The input for the tool are **.gb** files of known biosynthesis gene clusters (BGCs) responsible for production of specific group secoundary metabolite. The **clinker**, the package compares the BGCs and extracts gene groups from the alignment, thus having idea of how abundant specific gene is in number of input clusters. Thereafter the clusters are passed through **cblaster** to find similar BGCs to the input files. The tables that the tool creates are thereafter combined and narrowed down automatically to contain clusters that contain specific minimum number of core genes that are shared across all bgcs and specific number of tailoring genes. The data is then processed to be importable in **iTOL annotation edditor** for annotation of iTOL phylogenetic tree.
